cognify/optimizer/core/sh.py

```diff
@@ -0,0 +1,121 @@
+        sorted_indicator_indices = sorted(
+            not_converged,
+            key=lambda x: (-indicators[x][0], indicators[x][2], indicators[x][1])
+        )
+        runs_left_to_run = sorted_indicator_indices[:n_left]
+        self.ready_to_run = [self.ready_to_run[i] for i in runs_left_to_run]
+    
+    def _halving_jointly(self, indicators, not_converged, n_left):
+        """Round-Robin in each dimension
+        """
+        # NOTE: indicators all smaller the better
+        sorted_groups = list(map(sorted, zip(*indicators)))
+        # round robin select from top of each list
+        # avoid duplicates
+        runs_left_to_run = set()
+        from_group = 0
+        while len(runs_left_to_run) < n_left:
+            if sorted_groups[from_group]:
+                runs_left_to_run.add(sorted_groups[from_group].pop(0))
+            from_group = (from_group + 1) % len(sorted_groups)
+        self.ready_to_run = [self.ready_to_run[i] for i in runs_left_to_run]
+    
+    def run_and_prune(self):
+        for i in range(self.num_SH_iter):
+            if len(self.ready_to_run) == 0:
+                break
+            
+            # print(f"SH next with {self.ready_to_run}, budget: {self.initial_step_budget * self.prune_rate ** i}")
+            for j in self.ready_to_run:
+                self.num_inner_trials[j] += int(self.initial_step_budget * self.prune_rate ** i)
+                self.selected_runs[j].opt_config.n_trials = int(self.initial_step_budget * self.prune_rate ** i)
+                
+            futures: list[Future] = []
+            with ThreadPoolExecutor(max_workers=len(self.ready_to_run)) as executor:
+                for j in self.ready_to_run:
+                    futures.append(executor.submit(
+                        self._evaluate_one_config, j
+                    ))
+                    
+                try:
+                    outer_indicators = []
+                    not_converged = []
+                    for i, f in enumerate(futures):
+                        eval_result: EvaluationResult = f.result()
+                        # If next layer converged, even for promising ones, we will not continue them
+                        if not eval_result.meta.get("converged", False):
+                            not_converged.append(j)
+                        outer_indicators.append(CommonStats.objectives.select_from(-eval_result.reduced_score, eval_result.reduced_price, eval_result.reduced_exec_time))
+                except Exception as e:
+                    logger.error(f"Error in SH: {e}")
+                    raise
+            
+            # not consider converged ones
+            n =  int(len(self.selected_runs) * self.prune_rate ** -(i+1))
+            # self._halving_by_order(outer_indicators, not_converged, n)
+            self._halving_jointly(outer_indicators, not_converged, n)
+    
+    def execute(self) -> tuple[list[EvaluationResult], list[int]]:
```

Comment:
So when successive halving is turned on, is an instance of this class the driver for the optimization?

Reply:
right, it manages the optimization of a proposed config bucket